File: src/JetpacFR.Web/TraceCodecWeb.fs
namespace JetpacFR.Core

/// WEB SHELL SHIM (JetpacFR.Web): byte-compatible reimplementation of the
/// JetpacFR.Core TraceCodec (.jpt binary trace format) for the browser.
/// The Core module uses System.IO + Marshal/fixed (not Fable-compilable); this
/// one builds/parses the identical byte stream so traces saved in the browser
/// load in the desktop app and vice versa.
///
/// Format (all little-endian), verified against the desktop writer:
///   magic "JPTR" (4)
///   version u16 = 1 (2)
///   startTick u32, endTick u32 (8)
///   entryCount u32, snapCount u32, writeCount u32, portCount u32, frameCount u32 (20)
///   entries  entryCount x 20  (Pc u16 @0, B0..B3 u8 @2..5, Target u16 @6,
///                              Tick u32 @8, Length/Cycles/FlagsBefore/FlagsAfter/Taken u8 @12..16)
///   snaps    snapCount x 32   (Tick u32 @0, Af..Pc u16 @4..26, I/R u8 @28/29)
///   writes   writeCount x 8   (Tick u32 @0, Address u16 @4, OldValue u8 @6, NewValue u8 @7)
///   ports    portCount x 8    (Tick u32 @0, Port u16 @4, Value u8 @6, Border u8 @7)
///   frameTicks frameCount x 4
///   perPc    0x10000 x 4
///   selfMod  0x10000 (omitted by pre-self-mod writers; reader tolerates EOF)
module TraceCodecWeb =

    open System

    let private magic = [| byte 'J'; byte 'P'; byte 'T'; byte 'R' |]

    let private w8 (b: byte[]) (o: int) (v: int) = b[o] <- byte (v &&& 0xFF)

    let private w16 (b: byte[]) (o: int) (v: int) =
        b[o] <- byte (v &&& 0xFF)
        b[o + 1] <- byte ((v >>> 8) &&& 0xFF)

    let private w32 (b: byte[]) (o: int) (v: int) =
        b[o] <- byte (v &&& 0xFF)
        b[o + 1] <- byte ((v >>> 8) &&& 0xFF)
        b[o + 2] <- byte ((v >>> 16) &&& 0xFF)
        b[o + 3] <- byte ((v >>> 24) &&& 0xFF)

    let private r8 (b: byte[]) (o: int) = int b[o]
    let private r16 (b: byte[]) (o: int) = int b[o] ||| (int b[o + 1] <<< 8)

    let private r32 (b: byte[]) (o: int) =
        int b[o]
        ||| (int b[o + 1] <<< 8)
        ||| (int b[o + 2] <<< 16)
        ||| (int b[o + 3] <<< 24)

    /// Serialize a trace to the .jpt byte stream (identical layout to Core's writer).
    let encode (trace: Trace) : byte[] =
        let total =
            4
            + 2
            + 8
            + 20
            + trace.Entries.Length * 20
            + trace.Snapshots.Length * 32
            + trace.Writes.Length * 8
            + trace.Ports.Length * 8
            + trace.FrameTicks.Length * 4
            + 0x10000 * 4
            + 0x10000

        let b = Array.zeroCreate<byte> total
        let mutable o = 0
        Array.Copy(magic, 0, b, 0, 4)
        o <- o + 4
        w16 b o 1
        o <- o + 2
        w32 b o (int trace.StartTick)
        o <- o + 4
        w32 b o (int trace.EndTick)
        o <- o + 4
        w32 b o trace.Entries.Length
        o <- o + 4
        w32 b o trace.Snapshots.Length
        o <- o + 4
        w32 b o trace.Writes.Length
        o <- o + 4
        w32 b o trace.Ports.Length
        o <- o + 4
        w32 b o trace.FrameTicks.Length
        o <- o + 4

        for e in trace.Entries do
            w16 b o (int e.Pc)
            w8 b (o + 2) (int e.B0)
            w8 b (o + 3) (int e.B1)
            w8 b (o + 4) (int e.B2)
            w8 b (o + 5) (int e.B3)
            w16 b (o + 6) (int e.Target)
            w32 b (o + 8) (int e.Tick)
            w8 b (o + 12) (int e.Length)
            w8 b (o + 13) (int e.Cycles)
            w8 b (o + 14) (int e.FlagsBefore)
            w8 b (o + 15) (int e.FlagsAfter)
            w8 b (o + 16) (int e.Taken)
            o <- o + 20

        for s in trace.Snapshots do
            w32 b o (int s.Tick)
            w16 b (o + 4) (int s.Af)
            w16 b (o + 6) (int s.Bc)
            w16 b (o + 8) (int s.De)
            w16 b (o + 10) (int s.Hl)
            w16 b (o + 12) (int s.Af2)
            w16 b (o + 14) (int s.Bc2)
            w16 b (o + 16) (int s.De2)
            w16 b (o + 18) (int s.Hl2)
            w16 b (o + 20) (int s.Ix)
            w16 b (o + 22) (int s.Iy)
            w16 b (o + 24) (int s.Sp)
            w16 b (o + 26) (int s.Pc)
            w8 b (o + 28) (int s.I)
            w8 b (o + 29) (int s.R)
            o <- o + 32

        for w in trace.Writes do
            w32 b o (int w.Tick)
            w16 b (o + 4) (int w.Address)
            w8 b (o + 6) (int w.OldValue)
            w8 b (o + 7) (int w.NewValue)
            o <- o + 8

        for p in trace.Ports do
            w32 b o (int p.Tick)
            w16 b (o + 4) (int p.Port)
            w8 b (o + 6) (int p.Value)
            w8 b (o + 7) (int p.Border)
            o <- o + 8

        for t in trace.FrameTicks do
            w32 b o (int t)
            o <- o + 4

        for i in 0..0xFFFF do
            w32 b o trace.PerPcCount[i]
            o <- o + 4

        for i in 0..0xFFFF do
            w8 b o (if trace.SelfModified[i] then 1 else 0)
            o <- o + 1

        if o <> total then
            failwithf "TraceCodecWeb: wrote %d bytes, expected %d" o total

        b

    /// Parse a .jpt byte stream into a Trace (tolerates the omitted self-mod array).
    let decode (b: byte[]) : Trace =
        let mutable o = 0

        if b.Length < 34 then
            failwithf "trace file too short (%d bytes)" b.Length

        for i in 0..3 do
            if b[i] <> magic[i] then
                failwith "not a Jetpac trace file (bad magic)"

        o <- o + 4
        let v = r16 b o
        o <- o + 2

        if v <> 1 then
            failwithf "unsupported trace version %d" v

        let startTick = uint32 (r32 b o)
        o <- o + 4
        let endTick = uint32 (r32 b o)
        o <- o + 4
        let entryCount = r32 b o
        o <- o + 4
        let snapCount = r32 b o
        o <- o + 4
        let writeCount = r32 b o
        o <- o + 4
        let portCount = r32 b o
        o <- o + 4
        let frameCount = r32 b o
        o <- o + 4

        let need =
            entryCount * 20
            + snapCount * 32
            + writeCount * 8
            + portCount * 8
            + frameCount * 4
            + 0x10000 * 4

        if o + need > b.Length then
            failwithf "trace file truncated: header claims %d more bytes but only %d remain" need (b.Length - o)

        let entries = Array.zeroCreate<TraceEntry> entryCount

        for i in 0 .. entryCount - 1 do
            entries[i] <-
                { Pc = uint16 (r16 b o)
                  B0 = byte (r8 b (o + 2))
                  B1 = byte (r8 b (o + 3))
                  B2 = byte (r8 b (o + 4))
                  B3 = byte (r8 b (o + 5))
                  Target = uint16 (r16 b (o + 6))
                  Tick = uint32 (r32 b (o + 8))
                  Length = byte (r8 b (o + 12))
                  Cycles = byte (r8 b (o + 13))
                  FlagsBefore = byte (r8 b (o + 14))
                  FlagsAfter = byte (r8 b (o + 15))
                  Taken = byte (r8 b (o + 16)) }

            o <- o + 20

        let snaps = Array.zeroCreate<RegSnapshot> snapCount

        for i in 0 .. snapCount - 1 do
            snaps[i] <-
                { Tick = uint32 (r32 b o)
                  Af = uint16 (r16 b (o + 4))
                  Bc = uint16 (r16 b (o + 6))
                  De = uint16 (r16 b (o + 8))
                  Hl = uint16 (r16 b (o + 10))
                  Af2 = uint16 (r16 b (o + 12))
                  Bc2 = uint16 (r16 b (o + 14))
                  De2 = uint16 (r16 b (o + 16))
                  Hl2 = uint16 (r16 b (o + 18))
                  Ix = uint16 (r16 b (o + 20))
                  Iy = uint16 (r16 b (o + 22))
                  Sp = uint16 (r16 b (o + 24))
                  Pc = uint16 (r16 b (o + 26))
                  I = byte (r8 b (o + 28))
                  R = byte (r8 b (o + 29)) }

            o <- o + 32

        let writes = Array.zeroCreate<MemWriteEvent> writeCount

        for i in 0 .. writeCount - 1 do
            writes[i] <-
                { Tick = uint32 (r32 b o)
                  Address = uint16 (r16 b (o + 4))
                  OldValue = byte (r8 b (o + 6))
                  NewValue = byte (r8 b (o + 7)) }

            o <- o + 8

        let ports = Array.zeroCreate<PortEvent> portCount

        for i in 0 .. portCount - 1 do
            ports[i] <-
                { Tick = uint32 (r32 b o)
                  Port = uint16 (r16 b (o + 4))
                  Value = byte (r8 b (o + 6))
                  Border = byte (r8 b (o + 7)) }

            o <- o + 8

        let frameTicks = Array.zeroCreate<uint32> frameCount

        for i in 0 .. frameCount - 1 do
            frameTicks[i] <- uint32 (r32 b o)
            o <- o + 4

        let perPc = Array.zeroCreate<int> 0x10000

        for i in 0..0xFFFF do
            perPc[i] <- r32 b o
            o <- o + 4
        // Traces saved before self-mod tracking omit the final array; treat those
        // as having no self-modification evidence.
        let selfMod =
            if o + 0x10000 <= b.Length then
                let a = Array.zeroCreate<bool> 0x10000

                for i in 0..0xFFFF do
                    a[i] <- b[o + i] <> 0uy

                a
            else
                Array.zeroCreate<bool> 0x10000

        let firstAt = Array.create 0x10000 -1

        for i in 0 .. entries.Length - 1 do
            let pc = int entries[i].Pc

            if firstAt[pc] < 0 then
                firstAt[pc] <- i

        { Entries = entries
          Snapshots = snaps
          Writes = writes
          Ports = ports
          FrameTicks = frameTicks
          PerPcCount = perPc
          SelfModified = selfMod
          SelfModCount = selfMod |> Array.filter id |> Array.length
          FirstIndexAtPc = firstAt
          StartTick = startTick
          EndTick = endTick }

File: src/JetpacFR.Web/webroot/TraceCodecWeb.js

import { fill, item, setItem } from "./fable_modules/fable-library-js.5.17.0/Array.js";
import { Exception, copyToArray } from "./fable_modules/fable-library-js.5.17.0/Util.js";
import { printf, toFail } from "./fable_modules/fable-library-js.5.17.0/String.js";
import { Trace, PortEvent, MemWriteEvent, RegSnapshot, TraceEntry } from "./TraceTypesWeb.js";

const magic = new Uint8Array(["J".charCodeAt(0) & 0xFF, "P".charCodeAt(0) & 0xFF, "T".charCodeAt(0) & 0xFF, "R".charCodeAt(0) & 0xFF]);

function w8(b, o, v) {
    setItem(b, o, (v & 255) & 0xFF);
}

function w16(b, o, v) {
    setItem(b, o, (v & 255) & 0xFF);
    setItem(b, o + 1, ((v >> 8) & 255) & 0xFF);
}

function w32(b, o, v) {
    setItem(b, o, (v & 255) & 0xFF);
    setItem(b, o + 1, ((v >> 8) & 255) & 0xFF);
    setItem(b, o + 2, ((v >> 16) & 255) & 0xFF);
    setItem(b, o + 3, ((v >> 24) & 255) & 0xFF);
}

function r8(b, o) {
    return ~~item(o, b) | 0;
}

function r16(b, o) {
    return (~~item(o, b) | (~~item(o + 1, b) << 8)) | 0;
}

function r32(b, o) {
    return (((~~item(o, b) | (~~item(o + 1, b) << 8)) | (~~item(o + 2, b) << 16)) | (~~item(o + 3, b) << 24)) | 0;
}

/**
 * Serialize a trace to the .jpt byte stream (identical layout to Core's writer).
 */
export function encode(trace) {
    const total = (((((((34 + (trace.Entries.length * 20)) + (trace.Snapshots.length * 32)) + (trace.Writes.length * 8)) + (trace.Ports.length * 8)) + (trace.FrameTicks.length * 4)) + (65536 * 4)) + 65536) | 0;
    const b = new Uint8Array(total);
    let o = 0;
    copyToArray(magic, 0, b, 0, 4);
    o = ((o + 4) | 0);
    w16(b, o, 1);
    o = ((o + 2) | 0);
    w32(b, o, ~~trace.StartTick);
    o = ((o + 4) | 0);
    w32(b, o, ~~trace.EndTick);
    o = ((o + 4) | 0);
    w32(b, o, trace.Entries.length);
    o = ((o + 4) | 0);
    w32(b, o, trace.Snapshots.length);
    o = ((o + 4) | 0);
    w32(b, o, trace.Writes.length);
    o = ((o + 4) | 0);
    w32(b, o, trace.Ports.length);
    o = ((o + 4) | 0);
    w32(b, o, trace.FrameTicks.length);
    o = ((o + 4) | 0);
    const arr = trace.Entries;
    for (let idx = 0; idx <= (arr.length - 1); idx++) {
        const e = item(idx, arr);
        w16(b, o, ~~e.Pc);
        w8(b, o + 2, ~~e.B0);
        w8(b, o + 3, ~~e.B1);
        w8(b, o + 4, ~~e.B2);
        w8(b, o + 5, ~~e.B3);
        w16(b, o + 6, ~~e.Target);
        w32(b, o + 8, ~~e.Tick);
        w8(b, o + 12, ~~e.Length);
        w8(b, o + 13, ~~e.Cycles);
        w8(b, o + 14, ~~e.FlagsBefore);
        w8(b, o + 15, ~~e.FlagsAfter);
        w8(b, o + 16, ~~e.Taken);
        o = ((o + 20) | 0);
    }
    const arr_1 = trace.Snapshots;
    for (let idx_1 = 0; idx_1 <= (arr_1.length - 1); idx_1++) {
        const s = item(idx_1, arr_1);
        w32(b, o, ~~s.Tick);
        w16(b, o + 4, ~~s.Af);
        w16(b, o + 6, ~~s.Bc);
        w16(b, o + 8, ~~s.De);
        w16(b, o + 10, ~~s.Hl);
        w16(b, o + 12, ~~s.Af2);
        w16(b, o + 14, ~~s.Bc2);
        w16(b, o + 16, ~~s.De2);
        w16(b, o + 18, ~~s.Hl2);
        w16(b, o + 20, ~~s.Ix);
        w16(b, o + 22, ~~s.Iy);
        w16(b, o + 24, ~~s.Sp);
        w16(b, o + 26, ~~s.Pc);
        w8(b, o + 28, ~~s.I);
        w8(b, o + 29, ~~s.R);
        o = ((o + 32) | 0);
    }
    const arr_2 = trace.Writes;
    for (let idx_2 = 0; idx_2 <= (arr_2.length - 1); idx_2++) {
        const w = item(idx_2, arr_2);
        w32(b, o, ~~w.Tick);
        w16(b, o + 4, ~~w.Address);
        w8(b, o + 6, ~~w.OldValue);
        w8(b, o + 7, ~~w.NewValue);
        o = ((o + 8) | 0);
    }
    const arr_3 = trace.Ports;
    for (let idx_3 = 0; idx_3 <= (arr_3.length - 1); idx_3++) {
        const p = item(idx_3, arr_3);
        w32(b, o, ~~p.Tick);
        w16(b, o + 4, ~~p.Port);
        w8(b, o + 6, ~~p.Value);
        w8(b, o + 7, ~~p.Border);
        o = ((o + 8) | 0);
    }
    const arr_4 = trace.FrameTicks;
    for (let idx_4 = 0; idx_4 <= (arr_4.length - 1); idx_4++) {
        const t = item(idx_4, arr_4);
        w32(b, o, ~~t);
        o = ((o + 4) | 0);
    }
    for (let i = 0; i <= 65535; i++) {
        w32(b, o, item(i, trace.PerPcCount));
        o = ((o + 4) | 0);
    }
    for (let i_1 = 0; i_1 <= 65535; i_1++) {
        w8(b, o, item(i_1, trace.SelfModified) ? 1 : 0);
        o = ((o + 1) | 0);
    }
    if (o !== total) {
        const arg = o | 0;
        toFail(printf("TraceCodecWeb: wrote %d bytes, expected %d"))(arg)(total);
    }
    return b;
}

/**
 * Parse a .jpt byte stream into a Trace (tolerates the omitted self-mod array).
 */
export function decode(b) {
    let array_1;
    let o = 0;
    if (b.length < 34) {
        const arg = b.length | 0;
        toFail(printf("trace file too short (%d bytes)"))(arg);
    }
    for (let i = 0; i <= 3; i++) {
        if (item(i, b) !== item(i, magic)) {
            throw new Exception("not a Jetpac trace file (bad magic)");
        }
    }
    o = ((o + 4) | 0);
    const v = r16(b, o) | 0;
    o = ((o + 2) | 0);
    if (v !== 1) {
        toFail(printf("unsupported trace version %d"))(v);
    }
    const startTick = r32(b, o) >>> 0;
    o = ((o + 4) | 0);
    const endTick = r32(b, o) >>> 0;
    o = ((o + 4) | 0);
    const entryCount = r32(b, o) | 0;
    o = ((o + 4) | 0);
    const snapCount = r32(b, o) | 0;
    o = ((o + 4) | 0);
    const writeCount = r32(b, o) | 0;
    o = ((o + 4) | 0);
    const portCount = r32(b, o) | 0;
    o = ((o + 4) | 0);
    const frameCount = r32(b, o) | 0;
    o = ((o + 4) | 0);
    const need = ((((((entryCount * 20) + (snapCount * 32)) + (writeCount * 8)) + (portCount * 8)) + (frameCount * 4)) + (65536 * 4)) | 0;
    if ((o + need) > b.length) {
        const arg_3 = (b.length - o) | 0;
        toFail(printf("trace file truncated: header claims %d more bytes but only %d remain"))(need)(arg_3);
    }
    const entries = fill(new Array(entryCount), 0, entryCount, new TraceEntry(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0));
    for (let i_1 = 0; i_1 <= (entryCount - 1); i_1++) {
        setItem(entries, i_1, new TraceEntry(r16(b, o) & 0xFFFF, r8(b, o + 2) & 0xFF, r8(b, o + 3) & 0xFF, r8(b, o + 4) & 0xFF, r8(b, o + 5) & 0xFF, r16(b, o + 6) & 0xFFFF, r32(b, o + 8) >>> 0, r8(b, o + 12) & 0xFF, r8(b, o + 13) & 0xFF, r8(b, o + 14) & 0xFF, r8(b, o + 15) & 0xFF, r8(b, o + 16) & 0xFF));
        o = ((o + 20) | 0);
    }
    const snaps = fill(new Array(snapCount), 0, snapCount, new RegSnapshot(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0));
    for (let i_2 = 0; i_2 <= (snapCount - 1); i_2++) {
        setItem(snaps, i_2, new RegSnapshot(r32(b, o) >>> 0, r16(b, o + 4) & 0xFFFF, r16(b, o + 6) & 0xFFFF, r16(b, o + 8) & 0xFFFF, r16(b, o + 10) & 0xFFFF, r16(b, o + 12) & 0xFFFF, r16(b, o + 14) & 0xFFFF, r16(b, o + 16) & 0xFFFF, r16(b, o + 18) & 0xFFFF, r16(b, o + 20) & 0xFFFF, r16(b, o + 22) & 0xFFFF, r16(b, o + 24) & 0xFFFF, r16(b, o + 26) & 0xFFFF, r8(b, o + 28) & 0xFF, r8(b, o + 29) & 0xFF));
        o = ((o + 32) | 0);
    }
    const writes = fill(new Array(writeCount), 0, writeCount, new MemWriteEvent(0, 0, 0, 0));
    for (let i_3 = 0; i_3 <= (writeCount - 1); i_3++) {
        setItem(writes, i_3, new MemWriteEvent(r32(b, o) >>> 0, r16(b, o + 4) & 0xFFFF, r8(b, o + 6) & 0xFF, r8(b, o + 7) & 0xFF));
        o = ((o + 8) | 0);
    }
    const ports = fill(new Array(portCount), 0, portCount, new PortEvent(0, 0, 0, 0));
    for (let i_4 = 0; i_4 <= (portCount - 1); i_4++) {
        setItem(ports, i_4, new PortEvent(r32(b, o) >>> 0, r16(b, o + 4) & 0xFFFF, r8(b, o + 6) & 0xFF, r8(b, o + 7) & 0xFF));
        o = ((o + 8) | 0);
    }
    const frameTicks = new Uint32Array(frameCount);
    for (let i_5 = 0; i_5 <= (frameCount - 1); i_5++) {
        setItem(frameTicks, i_5, r32(b, o) >>> 0);
        o = ((o + 4) | 0);
    }
    const perPc = new Int32Array(65536);
    for (let i_6 = 0; i_6 <= 65535; i_6++) {
        setItem(perPc, i_6, r32(b, o) | 0);
        o = ((o + 4) | 0);
    }
    let selfMod;
    if ((o + 65536) <= b.length) {
        const a = fill(new Array(65536), 0, 65536, false);
        for (let i_7 = 0; i_7 <= 65535; i_7++) {
            setItem(a, i_7, item(o + i_7, b) !== 0);
        }
        selfMod = a;
    }
    else {
        selfMod = fill(new Array(65536), 0, 65536, false);
    }
    const firstAt = fill(new Int32Array(65536), 0, 65536, -1);
    for (let i_8 = 0; i_8 <= (entries.length - 1); i_8++) {
        const pc = ~~item(i_8, entries).Pc | 0;
        if (item(pc, firstAt) < 0) {
            setItem(firstAt, pc, i_8 | 0);
        }
    }
    return new Trace(entries, snaps, writes, ports, frameTicks, perPc, selfMod, (array_1 = selfMod.filter((x) => x), array_1.length), firstAt, startTick, endTick);
}

